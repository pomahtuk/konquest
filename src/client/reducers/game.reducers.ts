import {
  StateGameOptions,
  GameActionTypes,
  StartGameAction,
  AddPlayerTurnAction,
  SetGameOptionsAction,
  AddPlayerAction,
  SetPlanetAction,
  AddPlayerTurnOrderAction
} from "../actions/game.actions";
import ConquestGame, { GameStatus, TurnStatus } from "../../logic/Game";
import Player, { PlayerTurnOrder } from "../../logic/Player";
import Planet, { PlanetMap } from "../../logic/Planet";

export interface GameState {
  isStarted: boolean;
  status: GameStatus;
  winner: Player | null;
  activePlayer?: Player;
  players: Player[];
  planets: PlanetMap;
  gameOptions: StateGameOptions;
  gameStartError: boolean;
  turnError: boolean;
  errorText?: string;
  originPlanet?: Planet;
  destinationPlanet?: Planet;
  currentPlayerOrders: PlayerTurnOrder[];
  currentShipsModifier: { [key: string]: number };
}

const defaultState: GameState = {
  isStarted: false,
  status: GameStatus.NOT_STARTED,
  winner: null,
  players: [],
  planets: {},
  gameOptions: {
    fieldSize: 10,
    neutralPlanetCount: 4
  },
  gameStartError: false,
  turnError: false,
  currentPlayerOrders: [],
  currentShipsModifier: {}
};

let game: ConquestGame;

function konquestGame(
  state: GameState = defaultState,
  action: StartGameAction | AddPlayerTurnAction | SetGameOptionsAction | AddPlayerAction | SetPlanetAction | AddPlayerTurnOrderAction
): GameState {
  let turnStatus: TurnStatus;
  switch (action.type) {
    case GameActionTypes.SET_GAME_OPTIONS:
      return {
        ...state,
        gameOptions: {
          ...state.gameOptions,
          ...action.gameOptions
        }
      };
    case GameActionTypes.ADD_PLAYER:
      return {
        ...state,
        players: [...state.players, action.player]
      };
    case GameActionTypes.START_GAME:
      try {
        game = new ConquestGame({
          fieldHeight: state.gameOptions.fieldSize,
          fieldWidth: state.gameOptions.fieldSize,
          neutralPlanetCount: state.gameOptions.neutralPlanetCount,
          players: state.players.map((p) => new Player(p.screenName))
        });
        return {
          ...state,
          isStarted: true,
          gameStartError: false,
          status: game.status,
          winner: game.winner,
          players: game.getPlayers(),
          activePlayer: game.getPlayers()[game.waitingForPlayer],
          planets: game.getPlanets(),
          errorText: undefined,
          currentPlayerOrders: [],
          currentShipsModifier: {}
        };
      } catch (e) {
        return {
          ...state,
          gameStartError: true,
          errorText: e.message
        };
      }
    case GameActionTypes.ADD_PLAYER_TURN_ORDER:
      return {
        ...state,
        currentPlayerOrders: [...state.currentPlayerOrders, action.order],
        currentShipsModifier: [...state.currentPlayerOrders, action.order].reduce((acc, order): { [key: string]: number } => {
          acc[order.origin] = (acc[order.origin] || 0) + order.amount;
          return acc;
        }, {})
      };
    case GameActionTypes.ADD_PLAYER_TURN:
      if (game && state.isStarted === true) {
        turnStatus = game.addPlayerTurnData(action.turnData);
        if (turnStatus !== TurnStatus.INVALID) {
          return {
            ...state,
            status: game.status,
            winner: game.winner,
            players: game.getPlayers(),
            activePlayer: game.getPlayers()[game.waitingForPlayer],
            planets: game.getPlanets(),
            turnError: false,
            errorText: undefined,
            currentPlayerOrders: [],
            currentShipsModifier: {}
          };
        } else {
          return {
            ...state,
            turnError: true,
            errorText: "Turn data invalid"
          };
        }
      } else {
        return {
          ...state,
          turnError: true,
          errorText: "Game not started"
        };
      }
    case GameActionTypes.SET_ORIGIN_PLANET:
      return {
        ...state,
        originPlanet: action.planet
      };
    case GameActionTypes.SET_DESTINATION_PLANET:
      return {
        ...state,
        destinationPlanet: action.planet
      };
    default:
      return state;
  }
}

export default konquestGame;
