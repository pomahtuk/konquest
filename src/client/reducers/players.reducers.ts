import Player from "../../logic/Player";
import { AddPlayerAction, RemovePlayerAction, ClearPlayersAction, PlayerActionTypes } from "../actions/players.actions";

// const defaultState: Player[] = [new Player("Player One"), new ComputerPlayerEasy("Easy Computer")],
const defaultState: Player[] = [new Player("Player One"), new Player("Player Two")];

function players(state: Player[] = defaultState, action: AddPlayerAction | RemovePlayerAction | ClearPlayersAction): Player[] {
  switch (action.type) {
    case PlayerActionTypes.CLEAR_PLAYERS:
      return [];
    case PlayerActionTypes.ADD_PLAYER:
      return [...state, action.player];
    case PlayerActionTypes.REMOVE_PLAYER:
      return [...state.filter((p) => p.id !== action.player.id)];
    default:
      return state;
  }
}

export default players;
