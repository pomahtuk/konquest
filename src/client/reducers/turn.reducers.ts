import { PlayerTurnOrder } from "../../logic/Player";
import Planet from "../../logic/Planet";

import {
  SetPlanetAction,
  AddPlayerTurnOrderAction,
  TurnActionTypes,
  ClearTurnDataAction,
  RemovePlayerTurnOrderAction,
  UnsetPlanetsAction
} from "../actions/turn.actions";

export interface TurnStoreData {
  originPlanet?: Planet;
  destinationPlanet?: Planet;
  turnError: boolean;
  currentPlayerOrders: PlayerTurnOrder[];
  currentShipsModifier: { [key: string]: number };
}

const defaultState: TurnStoreData = {
  originPlanet: undefined,
  destinationPlanet: undefined,
  turnError: false,
  currentPlayerOrders: [],
  currentShipsModifier: {}
};

const calculateShipModifiers = (orderList: PlayerTurnOrder[]): { [key: string]: number } =>
  orderList.reduce((acc, order): { [key: string]: number } => {
    acc[order.origin] = (acc[order.origin] || 0) + order.amount;
    return acc;
  }, {});

function turn(
  state: TurnStoreData = defaultState,
  action: SetPlanetAction | AddPlayerTurnOrderAction | ClearTurnDataAction | RemovePlayerTurnOrderAction | UnsetPlanetsAction
): TurnStoreData {
  let orderList = [];
  switch (action.type) {
    case TurnActionTypes.SET_ORIGIN_PLANET:
      return {
        ...state,
        originPlanet: action.planet
      };
    case TurnActionTypes.SET_DESTINATION_PLANET:
      return {
        ...state,
        destinationPlanet: action.planet
      };
    case TurnActionTypes.ADD_PLAYER_TURN_ORDER:
      orderList = [...state.currentPlayerOrders, action.order];
      return {
        ...state,
        currentPlayerOrders: orderList,
        currentShipsModifier: calculateShipModifiers(orderList)
      };
    case TurnActionTypes.REMOVE_PLAYER_TURN_ORDER:
      orderList = [...state.currentPlayerOrders.slice(0, action.index), ...state.currentPlayerOrders.slice(action.index + 1)];
      return {
        ...state,
        currentPlayerOrders: orderList,
        currentShipsModifier: calculateShipModifiers(orderList)
      };
    case TurnActionTypes.CLEAR_TURN_DATA:
      return { ...defaultState };
    case TurnActionTypes.UNSET_PLANETS:
      return {
        ...state,
        originPlanet: undefined,
        destinationPlanet: undefined
      };
    default:
      return state;
  }
}

export default turn;
