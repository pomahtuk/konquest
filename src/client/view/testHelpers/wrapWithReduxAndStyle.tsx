import React, { ReactElement } from "react";
import { Provider } from "react-redux";
import storeCreator from "../../stores/game.store";
import ThemeProvider from "../themes/ThemeProvider";
import baseTheme from "../themes/base.theme";
const store = storeCreator();

export const CurrentStore = store;

const wrapWithReduxAndStyle = (element: ReactElement): ReactElement => {
  return (
    <Provider store={store}>
      <ThemeProvider theme={baseTheme}>{element}</ThemeProvider>
    </Provider>
  );
};

export default wrapWithReduxAndStyle;
