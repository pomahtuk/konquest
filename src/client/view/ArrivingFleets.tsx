import React, { ReactElement } from "react";
import { ArrivingFleet } from "../reducers/game.reducers";

export interface ArrivingFleetsProps {
  fleets: ArrivingFleet[];
}

const ArrivingFleets = ({ fleets }: ArrivingFleetsProps): ReactElement | null => {
  return fleets && fleets.length > 0 ? (
    <ul>
      {fleets.map((fleet) => (
        <li key={`${fleet.amount}-${fleet.destination}-${fleet.arrivingIn}`}>
          Fleet({fleet.amount}) arriving to {fleet.destination} in {fleet.arrivingIn} turns
        </li>
      ))}
    </ul>
  ) : null;
};

export default ArrivingFleets;
