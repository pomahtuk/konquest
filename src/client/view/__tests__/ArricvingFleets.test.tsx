import React from "react";
import { render } from "@testing-library/react";

import wrapWithReduxAndStyle from "../testHelpers/wrapWithReduxAndStyle";
import ArrivingFleets from "../ArrivingFleets";
import { ArrivingFleet } from "../../reducers/game.reducers";

describe("<ArrivingFleets />", (): void => {
  it("renders <ArrivingFleets /> component without any orders", (): void => {
    const { container } = render(wrapWithReduxAndStyle(<ArrivingFleets fleets={[]} />));
    expect(container).toMatchSnapshot();
  });

  it("renders <OrderList /> component with fleet", (): void => {
    const fleet: ArrivingFleet = { arrivingIn: 0, destination: "B", amount: 10, killPercent: 0.5 };
    const { container } = render(wrapWithReduxAndStyle(<ArrivingFleets fleets={[fleet]} />));
    expect(container).toMatchSnapshot();
  });
});
