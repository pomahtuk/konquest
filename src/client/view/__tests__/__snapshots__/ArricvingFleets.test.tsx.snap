// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ArrivingFleets /> renders <ArrivingFleets /> component without any orders 1`] = `<div />`;

exports[`<ArrivingFleets /> renders <OrderList /> component with fleet 1`] = `
<div>
  <ul>
    <li>
      Fleet(
      10
      ) arriving to 
      B
       in 
      0
       turns
    </li>
  </ul>
</div>
`;
